### DISCUSSION TITLE
Design: Why are array inputs to Block-Algorithms not const

### DISCUSSION THREAD

**Discussion by pauleonix** [NVIDIA/cccl#1288] [CUB] [answered]
I.e. why
```
__device__ __forceinline__ T 	cub::BlockReduce::Sum (T(&inputs)[ITEMS_PER_THREAD])
```
Instead of 
```
__device__ __forceinline__ T 	cub::BlockReduce::Sum (const T(&inputs)[ITEMS_PER_THREAD])
```
?
The way it is written right now makes const-correct code impossible.

**Comment by gevtushenko** [comment, marked as answer]
Hello @pauleonix! 

I agree, it should be const. I can speculate that the initial reason was to allow binary operator to accept non-const lvalues. We probably need something along the lines of std::reduce in our docs as well:

> The behavior is undefined if `binary_op` modifies any element

**Reply by pauleonix** [reply]
Should I open an issue for this?

**Reply by gevtushenko** [reply]
@pauleonix sure! This is an issue.
